workflow/artifacts/s3/s3.go

```diff
@@ -180,7 +180,27 @@ func (s3Driver *ArtifactDriver) Delete(artifact *wfv1.Artifact) error {
 		if err != nil {
 			return err
 		}
-		return s3cli.Delete(artifact.S3.Bucket, artifact.S3.Key)
+
+		isDir, err := s3cli.IsDirectory(artifact.S3.Bucket, artifact.S3.Key)
+		if err != nil {
+			return fmt.Errorf("failed to test if %s is a directory: %v", artifact.S3.Key, err)
+		}
+
+		if !isDir {
+			return s3cli.Delete(artifact.S3.Bucket, artifact.S3.Key)
+		} else {
+			keys, err := s3cli.ListDirectory(artifact.S3.Bucket, artifact.S3.Key)
+			if err != nil {
+				return fmt.Errorf("unable to list files in %s: %s", artifact.S3.Key, err)
+			}
+			for _, objKey := range keys {
+				err = s3cli.Delete(artifact.S3.Bucket, objKey)
```

Comment:
this can probably be parallelized.

I am also a bit concerned that for large directories this could potentially hit a rate limit or cause a lot of load, although I think in this case `archive: none` is already an edge case and a large dir would be another edge case, so potentially not necessary to think about that. though maybe should leave a comment for future reference

Reply:
`	err := retry.OnError(retry.DefaultBackoff, isTransientS3Err, func() error {`
currently we have `retry.DefaultBackoff` and `isTransientS3Err` to handle the rate limit situation

Reply:
Ah right above this code/surrounding it, good point.

I do think this could still be parallelized, although that can be a separate PR as it seems we lack parallelization in several (most?) places for artifacts (related: #12442)